cpp/test/util/reduction.cu

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../test_utils.cuh"
+
+#include <raft/random/rng.cuh>
+#include <raft/util/reduction.cuh>
+
+#include <rmm/device_scalar.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <numeric>
+
+namespace raft::util {
+
+template <typename ReduceLambda>
+__global__ void test_reduction_kernel(const int* input, int* reduction_res, ReduceLambda reduce_op)
+{
+  assert(gridDim.x == 1);
+  __shared__ int red_buf[raft::WarpSize];
```

Comment:
In order to allocate static shared memory without passing compile-time `blockDim.x`, I allocate it with the max size which is 32 warps worth (only Turing allows 64 warps which is an outlier, which we never test). But yeah, the correct way would be to get `blockDim.x` and allocate `blockDim.x/WarpSize` number of items.

Reply:
That makes sense. `raft::WarpSize` denotes the number of threads in a warp right? So it would be better to use a variable name something like `max_warps_per_block`.

Reply:
That's correct. Fixed it.